ENH: new GNSS TEC instrument

aburrell opened the pull request:
Adding a new instrument for vertical and slant TEC from MIT Haystack.

aburrell commented:
There's a problem listing files, which I don't know how to resolve.  Files download as expected, but the file list is not working.  Additionally, specifying a file to load does not work.  This needs to be fixed before I can check to see that the Datasets for the two data sets are formatted correctly.

jklenzing commented on pysatMadrigal/instruments/gnss_tec.py:
```diff
@@ -0,0 +1,216 @@
+
+# support list files routine
+# use the default pysat method
+dname = '{year:4d}{month:02d}{day:02d}'
```
I'm able to successfully download the files, but they use a two-digit year
```
gps171119g.002.hdf5
```

rstoneback replied:
There is also an issue in `ps_gen.list_files`. The two_digit_year_break keyword is not passed to the underlying from_os constructor. I have a working version locally.

`dname = '{year:2d}{month:02d}{day:02d}'`

aburrell replied:
Did not notice that.

rstoneback commented on pysatMadrigal/instruments/gnss_tec.py:
```diff
@@ -0,0 +1,216 @@
+supported_tags = {ss: {'vtec': "gps{:s}g{:s}.hdf5".format(dname, vname),
+                       'los': "los_{:s}{:s}.h5".format(dname, vname)}
+                  for ss in sat_ids.keys()}
+list_files = functools.partial(ps_gen.list_files,
```
```
list_files = functools.partial(ps_gen.list_files,
                               supported_tags=supported_tags,
                               two_digit_year_break=99)
```
Any two digit year less than 99 will be interpreted as 20XX

rstoneback commented on pysatMadrigal/instruments/gnss_tec.py:
```diff
@@ -0,0 +1,216 @@
+"""Supports the MIT Haystack GNSS TEC data products
+
+The Global Navigation Satellite System (GNSS) is used in conjunction with a
+world-winde receiver network to produce total electron content (TEC) data
```
`world-winde receiver`
minor typo
`world-wide receiver`

rstoneback commented on pysatMadrigal/instruments/gnss_tec.py:
```diff
@@ -0,0 +1,216 @@
+
+    """
+
+    ackn_str = ' '.join(["GPS TEC data products and access through the ",
```
There are spaces included in the text as well as a forced space ' ' with the join.

rstoneback commented on pysatMadrigal/instruments/gnss_tec.py:
```diff
@@ -0,0 +1,216 @@
+    password : string
+        Password for data download. (default=None)
+
+    Notes
```
```
Note
----
```

rstoneback commented:
I've started a pull on pysat with the bug fix for finding files.
pysat/#510
https://github.com/pysat/pysat/pull/510

rstoneback commented:
Loads data just fine.

```
In [8]: vtec.load(date=dt.datetime(2017, 11, 18))                                                                                                                                                                       

In [9]: vtec.data                                                                                                                                                                                                       
Out[9]: 
<xarray.Dataset>
Dimensions:   (gdalt: 1, gdlat: 180, glon: 360, time: 288)
Coordinates:
  * time      (time) datetime64[ns] 2017-11-18T00:02:30 ... 2017-11-18T23:57:30
  * gdlat     (gdlat) float64 -90.0 -89.0 -88.0 -87.0 ... 86.0 87.0 88.0 89.0
  * glon      (glon) float64 -180.0 -179.0 -178.0 -177.0 ... 177.0 178.0 179.0
  * gdalt     (gdalt) float64 350.0
Data variables:
    year      (time, gdlat, glon, gdalt) float64 nan nan nan nan ... nan nan nan
    month     (time, gdlat, glon, gdalt) float64 nan nan nan nan ... nan nan nan
    day       (time, gdlat, glon, gdalt) float64 nan nan nan nan ... nan nan nan
    hour      (time, gdlat, glon, gdalt) float64 nan nan nan nan ... nan nan nan
    min       (time, gdlat, glon, gdalt) float64 nan nan nan nan ... nan nan nan
    sec       (time, gdlat, glon, gdalt) float64 nan nan nan nan ... nan nan nan
    recno     (time, gdlat, glon, gdalt) float64 nan nan nan nan ... nan nan nan
    kindat    (time, gdlat, glon, gdalt) float64 nan nan nan nan ... nan nan nan
    kinst     (time, gdlat, glon, gdalt) float64 nan nan nan nan ... nan nan nan
    ut1_unix  (time, gdlat, glon, gdalt) float64 nan nan nan nan ... nan nan nan
    ut2_unix  (time, gdlat, glon, gdalt) float64 nan nan nan nan ... nan nan nan
    tec       (time, gdlat, glon, gdalt) float64 nan nan nan nan ... nan nan nan
    dtec      (time, gdlat, glon, gdalt) float64 nan nan nan nan ... nan nan nan
```

![gnss_example](https://user-images.githubusercontent.com/3595641/89950446-3f118180-dbef-11ea-9188-19df70435a2f.png)

aburrell commented:
@rstoneback can you provide your system and version details?  Because it won't load on mine.

aburrell commented:
Applied suggested changes and now the load works with the bugfix branch on pysat.  Currently working on:
- [ ] defining xarray with all variables except `tec` and `dtec` only depending on the time coordinate
- [ ] determining best way to load `los` data, and how to best use `sat_id` to define method.

For help with the second point, refer to: http://cedar.openmadrigal.org/static/siteSpecific/programming_tips.pdf

aburrell commented:
Moved the LOS task to issue #12.  Added a figure that begins to address issue #14.  Still need to fix one thing:

- [ ] Update metadata 'units' after loading, as they are incorrectly specified in the HDF file.

aburrell commented:
@rstoneback tried to figure out how to fix metadata, but failed miserably.  Advice?

rstoneback commented:
I'd change the units in the gnss_tec.py load method, after you get everything back from the general methods.

```
In [50]:  vtec.meta.data                                                                                                                                                                                                
Out[50]: 
              units long_name                                             desc     label      axis   scale notes value_min value_max fill
year              y      YEAR                            Year (universal time)      year      year  linear             NaN       NaN  NaN

In [51]: vtec.meta['year'] = {'units': 'new_units'}                                                                                                                                                                     

In [52]: vtec.meta.data                                                                                                                                                                                                 
Out[52]: 
              units long_name                                             desc     label      axis   scale notes value_min value_max fill
year      new_units      YEAR                            Year (universal time)      year      year  linear             NaN       NaN  NaN

```

I thought we could do something like, 
```
vtec.meta['year', 'units'] = 'new_units'
```
but I was wrong and I need to go fix some recent code.

rstoneback commented:
Correction for generality

 ```
 In [51]: vtec.meta['year'] = {vtec.units_label: 'new_units'}                                                                                                                                                                     
 
 ```

aburrell commented:
OK, this review is ready to convert pending pysat/pysat#510

rstoneback commented:
Test are running over in 510. Able to merge when those pass.

rstoneback commented:
> Test are running over in 510. Able to merge when those pass.

Pull was merged.

rstoneback commented on pysatMadrigal/instruments/gnss_tec.py:
```diff
@@ -0,0 +1,215 @@
+
+    # Fix the units for tec and dtec
+    if tag == 'vtec':
+        meta['tec'] = {'units': 'TECU'}
```
I messed up my generalization comment earlier. It works but not for this particular method.

The labels in Meta are set by the madrigal methods load. Technically the suggestion is more robust but we also control what the madrigal load returns.

rstoneback replied:
Things were posted out of order, suggestion referred to is a couple boxes below. with meta.units_label.

rstoneback commented on pysatMadrigal/instruments/methods/madrigal.py:
```diff
@@ -58,7 +56,13 @@ def load(fnames, tag=None, sat_id=None, xarray_coords=[]):
         List of keywords to use as coordinates if xarray output is desired
-        instead of a Pandas DataFrame (default=[])
+        instead of a Pandas DataFrame.  Can build an xarray Dataset with
+        that that have different coordinate dimensions by providing a dict
```
```suggestion
        that have different coordinate dimensions by providing a dict
```

rstoneback commented on pysatMadrigal/instruments/methods/madrigal.py:
```diff
@@ -58,7 +56,13 @@ def load(fnames, tag=None, sat_id=None, xarray_coords=[]):
     xarray_coords : list
         List of keywords to use as coordinates if xarray output is desired
-        instead of a Pandas DataFrame (default=[])
+        instead of a Pandas DataFrame.  Can build an xarray Dataset with
```
```suggestion
        instead of a Pandas DataFrame.  Can build an xarray Dataset with
```

rstoneback replied:
Hard for me to see what I changed with this one, I removed the 'with'

aburrell replied:
It didn't register in the suggestion.

rstoneback commented on pysatMadrigal/instruments/gnss_tec.py:
```diff
@@ -0,0 +1,215 @@
+
+    # Fix the units for tec and dtec
+    if tag == 'vtec':
+        meta['tec'] = {'units': 'TECU'}
```
```suggestion
        meta['tec'] = {meta.units_label: 'TECU'}
```

rstoneback commented on pysatMadrigal/instruments/gnss_tec.py:
```diff
@@ -0,0 +1,214 @@
+    # Fix the units for tec and dtec
+    if tag == 'vtec':
+        meta['tec'] = {meta.units_label: 'TECU'}
+        meta['dtec'] = {'units': 'TECU'}
```
```suggestion
        meta['dtec'] = {meta.units_label: 'TECU'}
```

aburrell commented:
The CI on this branch is failing because it takes too long to load a day of global VTEC data.  How do we want to address this?  Possibilities I have considered:

1. Make it possible to load specific ranges of data, including sub-day periods
2. Improve loading efficiency
3. Skip out of loading unit tests that take to long

aburrell commented:
Whilst fixing a bug I introduced when streamlining the HDF5 reading, I found some docs that suggest using netCDF over HDF with xarray.  It does seem to load faster, which may solve the erring problem.

aburrell commented:
@jklenzing any idea what's going on with Travis?? I am very confused.

jklenzing commented:
`mm_gen` needs to be updated to `ps_gen` on line 93 of the dmsp file.

aburrell commented:
I did it!  NetCDF solved the problem!  And I managed not to break the other instruments!!!

rstoneback commented on pysatMadrigal/instruments/gnss_tec.py:
```diff
@@ -0,0 +1,233 @@
+# use the default pysat method
+dname = '{year:02d}{month:02d}{day:02d}'
+vname = '.{version:03d}'
+supported_tags = {ss: {'vtec': "gps{:s}g{:s}.hdf5".format(dname, vname)}
```
This is hardcoded to .hdf5.

rstoneback commented on pysatMadrigal/instruments/methods/madrigal.py:
```diff
@@ -58,107 +54,217 @@ def load(fnames, tag=None, inst_id=None, xarray_coords=[]):
+                    # Recast as an xarray
+                    xdatasets.append(sel_data.to_xarray())
+
+                # Merge all of the datasets
+                for i in np.arange(1, len(xdatasets)):
+                    xdatasets[0] = xdatasets[0].merge(xdatasets[i])
```
```suggestion
                # Merge all of the datasets
                if len(xdatasets) > 1:
                    xdatasets = xr.merge(xdatasets, dim='time')
                else:
                    xdatasets = xdatasets[0]
```

I wasn't able to get hdf5 support going so I haven't been able to directly test this.

rstoneback commented on pysatMadrigal/instruments/gnss_tec.py:
```diff
@@ -0,0 +1,233 @@
+        Password for data download. (default=None)
+    url : string
+        URL for Madrigal site (default='http://cedar.openmadrigal.org')
+    file_format : string
```
The `file_format` keyword conflicts with the built-in keyword at the Instrument level. I remember you said something about this during the meeting today but unfortunately I don't remember what. I'm having trouble in this area.

I tried checking out 'hdf5' support. Using the branch over in pysat (without the keyword) I get,
```
gps = pysat.Instrument('gnss', 'tec', tag='vtec')                                                                                                                                                                   
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-8-c7e28318904d> in <module>
----> 1 gps = pysat.Instrument('gnss', 'tec', tag='vtec')

~/Code/pysat/pysat/_instrument.py in _get_supported_keywords(local_func)
   2950         for pop in pop_list[::-1]:
   2951             args.pop(pop)
-> 2952             defaults.pop(pop)
   2953 
   2954     out_dict = {}

IndexError: pop index out of range

```

With the keyword,
```
In [9]: gps = pysat.Instrument('gnss', 'tec', tag='vtec', file_format='hdf5')                                                                                                                                               
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-06a989677482> in <module>
----> 1 gps = pysat.Instrument('gnss', 'tec', tag='vtec', file_format='hdf5')

~/Code/pysat/pysat/_instrument.py in __init__(self, platform, name, tag, inst_id, clean_level, update_files, pad, orbit_info, inst_module, multi_file_day, manual_org, directory_format, file_format, temporary_file_list, strict_time_flag, ignore_empty_files, units_label, name_label, notes_label, desc_label, plot_label, axis_label, scale_label, min_label, max_label, fill_label, **kwargs)
    252                 estr = 'file format set to default, supplied string must be '
    253                 estr = '{:s}iteratable [{:}]'.format(estr, self.file_format)
--> 254                 raise ValueError(estr)
    255 
    256         # set up empty data and metadata

ValueError: file format set to default, supplied string must be iteratable [hdf5]
```

rstoneback replied:
I was using the wrong pysat branch. You had the correct link above. My bad. Trying again...

rstoneback replied:
My test setup was good the first time. Clearly I got myself turned around a bit.

The test for 'hdf5' was done with the 'delimiter_bug` branch. I confirmed that I get the pop issue or the ValueError.

aburrell replied:
Yeah, I apparently changed this bug in the following PR.  Will update it here, too.  Problems of finding too many related problems...

rstoneback commented on pysatMadrigal/instruments/gnss_tec.py:
```diff
@@ -0,0 +1,233 @@
+    return
+
+
+def download(date_array, tag='', inst_id='', data_path=None, user=None,
```
I downloaded a few days for testing but the files I got are labeled `stuff.hdf`. The load time is really quick though, suggesting they are actually netcdf files. Poking at the code the 'netcdf' path is indeed being run.

aburrell replied:
yeah, regardless of the format you ask for and get, the remote files are listed as .hdf5.  Not sure if I fixed the rename here or in the other PR.

rstoneback commented on pysatMadrigal/instruments/gnss_tec.py:
```diff
@@ -0,0 +1,233 @@
+        Tag identifier used for particular dataset. This input is provided by
+        pysat. (default='')
+    inst_id : string
+        Satellite ID string identifier used for particular dataset. This input
```
```suggestion
        Instrument ID string identifier used for particular dataset. This input
```

rstoneback reviewed (commented):
Thanks @aburrell! Comments posted. I ran into some issues so I wasn't able to test the hdf support. 

I was able to load multiple days (netcdf files) using the appropriate pysat branch for multiple days. Thanks for the additional improvements!

rstoneback commented on pysatMadrigal/instruments/methods/madrigal.py:
```diff
@@ -58,107 +54,217 @@ def load(fnames, tag=None, inst_id=None, xarray_coords=[]):
+
+                # Test to see that all data was retrieved
+                test_variables = [xkey for xkey
+                                  in xdatasets[0].variables.keys()]
```
```suggestion
                                  in xdatasets.variables.keys()]
```

rstoneback commented on pysatMadrigal/instruments/methods/madrigal.py:
```diff
@@ -58,107 +54,217 @@ def load(fnames, tag=None, inst_id=None, xarray_coords=[]):
+                                                  '{:d}; '.format(ldata),
+                                                  estr]))
+
+                data = xdatasets[0]
```
```suggestion
                data = xdatasets
```

rstoneback reviewed (commented):
I forgot a couple consistency suggestions to go along with one of my inputs.

aburrell commented on pysatMadrigal/instruments/gnss_tec.py:
```diff
@@ -0,0 +1,233 @@
+        Path to directory to download data to. (default=None)
+    user : string
+        User string input used for download. Provided by user and passed via
+        pysat. If an account is required for dowloads this routine here must
+        error if user not supplied. (default=None)
```
```suggestion
        pysat. (default=None)
```

aburrell commented on pysatMadrigal/instruments/methods/madrigal.py:
```diff
@@ -58,107 +54,217 @@ def load(fnames, tag=None, inst_id=None, xarray_coords=[]):
+    for fname in fnames:
+        if file_format == "netcdf4":
+            # Xarray natively opens netCDF data into a Dataset
+            file_data = xr.open_dataset(fname)
```
That would be better.

aburrell commented on pysatMadrigal/instruments/methods/madrigal.py:
```diff
@@ -58,107 +54,217 @@ def load(fnames, tag=None, inst_id=None, xarray_coords=[]):
+                                                 'desc': desc_string}
+
+            # Add an index
+            file_data = file_data.rename({'timestamps': 'time'})
+            data = file_data.set_index(time='time')
```
So, this was fixed in the following PR. >_<

aburrell commented on pysatMadrigal/instruments/methods/madrigal.py:
```diff
@@ -58,107 +54,217 @@ def load(fnames, tag=None, inst_id=None, xarray_coords=[]):
+        fdata.append(data)
+
+    # If multiple files were loaded, merge the data together
+    for i in np.arange(1, len(fdata)):
+        fdata[0] = fdata[0].merge(fdata[i])
+    data = fdata[0]
```
It doesn't ring a bell, but that doesn't mean it's not there.  Could you provide a link to the recommendation?  The merging/concating stuff is always a pain.

aburrell commented:
This instrument will now be added with the merge of #20
